rfcs/0005-range-queries.md

```diff
@@ -0,0 +1,256 @@
+    }
+
+}
+
+struct DbRecord {
+    key: Bytes,
+    value: Bytes,
+}
+```
+
+Note that the `DbIterator` may be invalidated internally if any of its referenced
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`
+    pub async fn scan_with_options<T>(
+        &self,
+        range: &T
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError> where
+        T: RangeBounds<Bytes> {
+	...
+    }
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes tables in memory which which have not been flushed and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables. The latest records will have precedence following
+the same rules that a normal `get` query follows.
+
+We will rely on the latest checkpoint state in order to obtain references to memtables, SSTs,
+and sorted runs needed for the query. The `DbIterator` will expose the UUID of the checkpoint
+that it was generated from. The `DbIterator` will only maintain references to
+tables which cover the range specified in the `scan` parameters.
+
+Note that if the `DbReader` was created with a reference to a specific checkpoint,
+then the `DbIterator` will only refer to that checkpoint.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. We will build an iterator which sorts
+and merges memtables (e.g. with `MemTableIterator`), L0 tables (`SstIterator`),
+and sorted runs (`SortedRunIterator`). 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the query began scanning its range of records. For example,
+a query may include only a subset of records that were inserted in the same 
```

Comment:
Hmm. Is this true? the `batch_write` code grabs a write guard on DB state and doesn't release it until the batch is merged.

Reply:
I guess it depends on our implementation. If the `DbIterator` references an iterator over the current memtable through the checkpoint (`DbStateSnapshot`), then I'm not sure how we could avoid it without holding the read lock for the duration of the scan. In my prototype, I pre-merged these records, so it would be possible to do all of that under the read lock. However, I thought that the benefit of using crossbeam's `SkipMap` was to allow iteration with concurrent writes.